src/lib/comms/sol-socket-impl-contiki.c

```diff
@@ -0,0 +1,297 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2016 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+
+#include <contiki-net.h>
+
+#include "sol-event-handler-contiki.h"
+#include "sol-log.h"
+#include "sol-mainloop.h"
+#include "sol-socket-impl.h"
+#include "sol-network-util.h"
+#include "sol-vector.h"
+
+struct sol_socket_contiki {
+    struct sol_socket base;
+
+    bool (*on_can_read)(void *data, struct sol_socket *s);
+    bool (*on_can_write)(void *data, struct sol_socket *s);
+    const void *data;
+
+    struct simple_udp_connection udpconn;
+
+    struct sol_ptr_vector pending_buffers;
+    struct sol_timeout *write_timeout;
+
+    bool read_monitor;
+    bool write_monitor;
+    bool event_posted;
+};
+
+struct pending_buffer {
+    struct sol_network_link_addr addr;
+    uint16_t datalen;
+    uint8_t data[];
+};
+
+extern struct process soletta_app_process;
+process_event_t sol_udp_socket_event;
+static bool registered_process = false;
+
+#define container_of(ptr, type, member) \
+    (type *)((char *)(ptr) - offsetof(type, member))
+
+static void
+receive_cb(struct simple_udp_connection *conn, const uip_ipaddr_t *srcaddr,
+    uint16_t srcport, const uip_ipaddr_t *dstaddr, uint16_t dstport,
+    const uint8_t *data, uint16_t datalen)
+{
+    struct sol_socket_contiki *socket;
+    struct pending_buffer *buf;
+    int ret;
+
+    socket = container_of(conn, struct sol_socket_contiki, udpconn);
+
+    buf = malloc(sizeof(*buf) + datalen);
+    SOL_NULL_CHECK(buf);
+
+    buf->addr.family = SOL_NETWORK_FAMILY_INET6;
+    memcpy(&buf->addr.addr.in6, srcaddr, sizeof(buf->addr.addr.in6));
+    buf->addr.port = srcport;
+
+    buf->datalen = datalen;
+    memcpy(buf->data, data, datalen);
+
+    ret = sol_ptr_vector_append(&socket->pending_buffers, buf);
+    SOL_INT_CHECK_GOTO(ret, < 0, append_error);
+
+    if (socket->event_posted)
```

Comment:
On Tue, 19 Apr 2016 06:31:58 -0700, Vinicius Gomes wrote:
> > +    socket = container_of(conn, struct sol_socket_contiki, udpconn);
> > +
> > +    buf = malloc(sizeof(*buf) + datalen);
> > +    SOL_NULL_CHECK(buf);
> > +
> > +    buf->addr.family = SOL_NETWORK_FAMILY_INET6;
> > +    memcpy(&buf->addr.addr.in6, srcaddr, sizeof(buf->addr.addr.in6));
> > +    buf->addr.port = srcport;
> > +
> > +    buf->datalen = datalen;
> > +    memcpy(buf->data, data, datalen);
> > +
> > +    ret = sol_ptr_vector_append(&socket->pending_buffers, buf);
> > +    SOL_INT_CHECK_GOTO(ret, < 0, append_error);
> > +
> > +    if (socket->event_posted)
> 
> [optional] I'm wondering if you could use the length of `pending_buffers` instead of adding this field: you only post the event if the length of pending buffers is 1.

Contiki's protothreads messing with my notion of concurrency...